src/relocate.rs

```diff
@@ -213,6 +227,10 @@ pub(super) unsafe fn relocate(envp: *mut *mut u8) {
         }
     }
 
+    // FIXME split function into two here with a hint::black_box around the
+    // function pointer to prevent the compiler from moving code between the
+    // functions.
```

Comment:
I've thought about something like this too, however I think it's actually better if we don't have to do this. If it's dangerous for this code at the end of `relocate` to mix with our relocation code, that suggests our relocation code itself is dangerous.

So instead, we use inline asm to do the actual relocation loads and stores, which hides the crucial part which Rust can't think about from Rust, and it should also have the effect of obliging the optimizer to preserve the ordering of the relocations with other code.

Reply:
The inline asm doesn't prevent moving a load of a static that needs relocations to before the inline asm as the compiler assumes that relocations have already happened and loading a static doesn't have any side effects. Musl does this kind of fence between every relocation step (after `R_RELATIVE` relocation resolving (`_dlstart_c` -> `__dls2`), after setting the thread pointer (`__dls2b` -> `__dls3`) and finally after symbol relocations resolving and right before calling the entry point of the main program (`__dls3` -> `AT_ENTRY`).

Reply:
I guess on a practical level, you're right that a `black_box` would avoid some obscure possibilities, though not all. I'd accept a PR for it.

On a theoretical level though, Rust doesn't make the kinds of guarantees that this code ultimately needs, and `black_box`, function pointers, volatile, `asm!` barriers, etc. aren't enough. We'd need guarantees like:

 - "The compiler shall not codegen certain language constructs into calls
   to other crates, such as translating `x == null()` to `x.is_null()`.
 - "The compiler shall not codegen certain language constructs using uses
   of static data."

Rust doesn't make these kinds of guarantees. C doesn't either, even with optimization barriers. My best idea right now for making this code theoretically sound is to precompile `relocate`, hand-verify it, and embed it with `include_bytes` in the `_start` code so that it runs before any Rust code runs.